src/groovy/org/transmartproject/db/dataquery/highdim/snp_lz/TokenList.groovy

```diff
@@ -0,0 +1,133 @@
+package org.transmartproject.db.dataquery.highdim.snp_lz
+
+import java.io.IOException;
+import java.io.Reader;
+import java.lang.reflect.UndeclaredThrowableException;
+import java.util.InputMismatchException;
+
+/**
+ * TokenList
+ *
+ * A wrapper around a StringBuilder, storing a sequence of characters,
+ * and an array of end indices, that together encode a list of Strings:
+ * element <var>i</var> of the list is stored in the StringBuilder
+ * at positions <code>endIndexes[i-1]</code> till <code>endIndexes[i]</code>
+ * (where <code>endIndexes[-1]</code> is 0 and <code>endIndexes</code> is 
+ * undefined for indices &ge; <var>expectedSize</var>).
+ * Empty values are supported.
+ *
+ * It comes with a {@link #parse(Reader)} function that parses a space-delimited
+ * stream of values.
+ *
+ * @author gijs@thehyve.nl
+ */
+class TokenList extends AbstractList<String> {
+
+    private StringBuilder builder = new StringBuilder()
+    private int size = 0
+    private int expectedSize
+    private int[] endIndexes
+    private char[] buffer = new char[1]
+
+    /**
+     * Initialises the TokenList. Sets the expected number of tokens.
+     * The TokenList does not support more tokens. The parser function 
+     * {@link #parse(Reader)} will throw an error if less tokens can be read.
+     *
+     * @param expectedSize the expected number of expected tokens.
+     */
+    public TokenList(int expectedSize) {
+        this.expectedSize = expectedSize
+        this.endIndexes = new int[expectedSize]
+    }
+
+    public int getExpectedSize() {
+        return expectedSize;
+    }
+
+    /**
+     * Reads a space-delimited stream of tokens from a Reader. If more or less than
+     * <var>expectedSize</var> tokens are read, an {@link InputMismatchException}
+     * is thrown.
+     *
+     * Empty tokens are supported: two consecutive spaces is interpreted as an empty token.
+     * Empty input is interpreted as an empty list, not as a singleton list with an empty token.
+     *
+     * @param r the reader.
+     */
+    public void parse(Reader r) {
+        try {
+            int n
+            char c
+            while ((n = r.read(buffer)) > 0) {
+                c = buffer[0]
+                if (c == ' ') {
+                    commitToken()
+                } else {
+                    addChar(c)
+                }
+            }
+            commitTokenIfInputNotEmpty()
+            if (this.size() < expectedSize) {
+                throw new InputMismatchException("Expected " +
+                        expectedSize + " tokens, but got only " +
+                        String.valueOf(this.size()));
+            }
+        } catch (IOException e) {
+            throw new UndeclaredThrowableException(e);
+        }
```

Comment:
This is a Groovy file. `UndeclaredThrowableException` is not needed. Remove also the semi-colons (also on the import list) and the unneeded `return`s.